yt/frontends/amrvac/data_structures.py

```diff
@@ -250,15 +251,110 @@ def _set_code_unit_attributes(self):
+        setdefaultattr(self, "density_unit", amrvac_units_dict['density_unit'])
+        setdefaultattr(self, "numberdensity_unit", amrvac_units_dict['numberdensity_unit'])
+
+        setdefaultattr(self, "temperature_unit", amrvac_units_dict['temperature_unit'])
+        setdefaultattr(self, "pressure_unit", amrvac_units_dict['pressure_unit'])
+        setdefaultattr(self, "magnetic_unit", amrvac_units_dict['magnetic_unit'])
+
+
+    def _get_amrvac_normalisations(self):
+        # frontend specific method
+        # note: normalisations in AMRVAC have 3 degrees of freedom. The user can specify a unit length and unit
+        # numberdensity, with the third option either a unit temperature OR unit velocity.
+        # If unit temperature is specified then unit velocity will be calculated accordingly, and vice-versa.
+        # AMRVAC does not allow to specify any other normalisation beside those four.
+
+        # note: _override_code_units() has already been called, so self.units_override has been set.
+        # _override_code_units() sets supplied items in units_override as attributes.
+        # Handled items are length, time, mass, velocity, magnetic, temperature
+        self._check_override_consistency()
+
+        # note: yt sets hydrogen mass equal to proton mass, amrvac doesn't.
+        mp_cgs = self.quan(1.672621898e-24, 'g')       # This value is taken from AstroPy
+        He_abundance = 0.1  # hardcoded parameter in AMRVAC
+
+        # get self.length_unit if overrides are supplied, otherwise use default
+        length_unit = getattr(self, 'length_unit', self.quan(1, 'cm'))
+
+        # 1. calculations for mass, density, numberdensity
+        if self.units_override.get('mass_unit') is not None:
+            # in this case unit_mass is supplied (and has been set as attribute)
+            mass_unit = self.mass_unit
+            density_unit = mass_unit / length_unit ** 3
+            numberdensity_unit = density_unit / ((1.0 + 4.0 * He_abundance) * mp_cgs)
+        else:
+            # other case: numberdensity is supplied. Fall back to one (default) if no overrides supplied
+            numberdensity_override = self.units_override.get('numberdensity_unit', (1, 'cm**-3'))
+            numberdensity_unit = self.quan(*numberdensity_override) # numberdensity is never set as attribute
+            density_unit = (1.0 + 4.0 * He_abundance) * mp_cgs * numberdensity_unit
+            mass_unit = density_unit * length_unit ** 3
+
+        # 2. calculations for velocity
+        if self.units_override.get('time_unit') is not None:
+            # in this case time was supplied
+            velocity_unit = length_unit / self.time_unit
+        else:
+            # other case: velocity was supplied. Fall back to zero (default) if no overrides supplied
+            velocity_unit = getattr(self, 'velocity_unit', self.quan(0, 'cm*s**-1'))
+
+        # 3. calculations for pressure and temperature
+        if velocity_unit.value != 0:
```

Comment:
I really don't see the point of using `self.quan(0, 'cm*s**-1')` as a default here. `None` would do the job equally well and be much easier to understand :)
(for reference, I attempted 5 times to provide a constructive proposal for these lines in my confusion, now I think I got it)

Reply:
Yeah you're right, I was thinking too much like it was implemented in amrvac. Changed this to None, much easier to read now